library/core/src/ptr/mod.rs

```diff
@@ -437,23 +437,43 @@ mod mut_ptr;
 ///
 /// # Safety
 ///
-/// Behavior is undefined if any of the following conditions are violated:
+/// Immediately upon executing, `drop_in_place` takes out a mutable borrow on the
+/// pointed-to-value. Effectively, this function is implemented like so:
+///
+/// ```
+/// # struct Foo { x: i32 }
+/// unsafe fn drop_in_place(to_drop: *mut Foo) {
+///     drop_in_place_inner(&mut *to_drop);
+///     unsafe fn drop_in_place_inner(to_drop: &mut Foo) {
+///         // ... drop the fields of `value` ...
+///     }
+/// }
+/// ```
+///
+/// This implies that the behavior is undefined if any of the following
+/// conditions are violated:
 ///
 /// * `to_drop` must be [valid] for both reads and writes.
 ///
-/// * `to_drop` must be properly aligned.
+/// * `to_drop` must be properly aligned, even if `T` has size 0.
+///
+/// * `to_drop` must be nonnull, even if `T` has size 0.
+///
+/// * The value `to_drop` points to must be valid for dropping, which may mean
```

Comment:
This particular requirement is ill-placed in the list since it is not a consequence of the things said above the enumeration -- but the docs claim that all list items follow from that prior information.

I think it is better to just state the requirement, we don't need to document hiw they come about because from the internal implementation details.

Reply:
Makes sense, undid the change to the introduction of this section